### PR TITLE
WIP

### LINKED ISSUE #844: Quest - Add UI for Tasks
**Name of feature to enhance**
Add UI to be able to view Tasks 

**Description of feature**
Add UI with Tasks. Labels can be either text or icons 

**Describe the enhancement**
See screenshots 

Screen Shot 2021-12-07 at 12 23 42 PM

Link to icons  
1. Overdue - https://drive.google.com/file/d/10RZSKD-JAgsbSoNcfVZiDYGQplz38BF8/view?usp=sharing
2. Due - https://drive.google.com/file/d/1Y5OiTYwkle5Izk_Bn58QMhyfU8gZ7UvS/view?usp=sharing

**Additional context** 
Add support for both labels and Icons  - controlled vis configuration settings we can support both and other alternatives

**Pre-Req task** 
- #857

### PR DESCRIPTION
**Pull request by FikriMilano**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Fixes #844

**Type**
Choose one: Feature

**Checklist**
- [ ] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [ ] I have added any strings visible on UI components to the `strings.xml` file
- [ ] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [ ] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [ ] I have built and run the fhircore app to verify my change fixes the issue and/or does not break the app

### CODE REVIEW

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/PatientTaskFragment.kt [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+          data.name.contains(value.toString(), ignoreCase = true) ||
+            data.id.contentEquals(value.toString())
+      }
+      RegisterFilterType.OVERDUE_FILTER -> {
+        data.overdue
+      }
+    }
+  }
+
+  @Suppress("UNCHECKED_CAST")
```
Comment (pld) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/PatientTaskFragment.kt:94]: can we check the type before we cast and remove this suppression?

**Review by pld — CHANGES_REQUESTED** [review]

File: android/engine/src/main/java/org/smartregister/fhircore/engine/ui/components/PaginatedRegisterViews.kt [hunk 2]
```diff
@@ -282,6 +294,7 @@ fun PaginatedRegisterPreviewWithResults() {
     showResultsCount = true,
     resultCount = 0,
     body = { Text(text = "Something cool") },
+    showPageCount = true,
     currentPage = 0,
     pagesCount = 20,
```
[the remark this answers is not in the document]
Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/ui/components/PaginatedRegisterViews.kt:299]: You mean the pagesCount?

File: android/engine/src/main/java/org/smartregister/fhircore/engine/ui/components/PaginatedRegisterViews.kt [hunk 3]
```diff
@@ -298,6 +311,7 @@ fun PaginatedRegisterPreviewWithoutResults() {
     showResultsCount = false,
     resultCount = 0,
     body = { Text(text = "Something cool") },
+    showPageCount = false,
     currentPage = 0,
     pagesCount = 20,
```
[the remark this answers is not in the document]
Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/ui/components/PaginatedRegisterViews.kt:316]: sure

File: android/engine/src/main/java/org/smartregister/fhircore/engine/ui/components/PaginatedRegisterViews.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (pld) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/ui/components/PaginatedRegisterViews.kt:299]: yep

File: android/quest/src/main/java/org/smartregister/fhircore/quest/data/task/PatientTaskRepository.kt [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        val patientId = task.`for`.reference.replace("Patient/", "")
+        val patient = fhirEngine.load(Patient::class.java, patientId)
+
+        domainMapper.mapToDomainModel(PatientTask(patient, task))
+      }
+    }
+  }
+
+  override suspend fun countAll(): Long {
+    return fhirEngine.count<Task> { filterByPractitioner(Task.OWNER, "6744") }
```
[the remark this answers is not in the document]
Comment (FikriMilano) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/data/task/PatientTaskRepository.kt:62]: Replied via this comment: https://github.com/opensrp/fhircore/pull/1023/files/981bfddb697708243cf43325a0eaf6dc64c3d369#r807480052

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/PatientTaskFragment.kt [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    super.onViewCreated(view, savedInstanceState)
+    registerViewConfiguration =
+      configurationRegistry.retrieveConfiguration<RegisterViewConfiguration>(
+        configClassification = QuestConfigClassification.PATIENT_TASK_REGISTER,
+      )
+    configureViews(registerViewConfiguration)
+  }
+
+  override fun navigateToDetails(uniqueIdentifier: String) {
+    // TODO: Nothing to navigate at the moment
```
[the remark this answers is not in the document]
Comment (FikriMilano) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/PatientTaskFragment.kt:55]: Yeah :D

File: android/quest/src/main/java/org/smartregister/fhircore/quest/data/task/PatientTaskRepository.kt [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  private val dispatcherProvider: DispatcherProvider
+) : RegisterRepository<PatientTask, PatientTaskItem> {
+
+  override suspend fun loadData(
+    query: String,
+    pageNumber: Int,
+    loadAll: Boolean
+  ): List<PatientTaskItem> {
+    return withContext(dispatcherProvider.io()) {
+      val tasks = fhirEngine.search<Task> { filterByPractitioner(Task.OWNER, "6744") }
```
[the remark this answers is not in the document]
Comment (pld) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/data/task/PatientTaskRepository.kt:50]: Hmm, OK, cool. What will this look like when that issue is closed?

Won't this ID be the ID of the currently logged in user, isn't that who we are showing tasks for?

File: android/engine/src/main/java/org/smartregister/fhircore/engine/ui/components/PaginatedRegisterViews.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (FikriMilano) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/ui/components/PaginatedRegisterViews.kt:299]: okay

File: android/quest/src/main/java/org/smartregister/fhircore/quest/data/task/PatientTaskRepository.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (FikriMilano) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/data/task/PatientTaskRepository.kt:50]: yeah, exactly. It will use the ID of the currently logged in user

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/PatientTaskFragment.kt [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+      clickListener = { listenerIntent, data -> onItemClicked(listenerIntent, data) }
+    )
+  }
+
+  override fun onItemClicked(listenerIntent: ListenerIntent, data: PatientTaskItem) {
+    // Nothing to navigate at the moment
+  }
+
+  // todo maybe we need to do a db call
+  override fun performFilter(
```
[the remark this answers is not in the document]
Comment (FikriMilano) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/PatientTaskFragment.kt:76]: sure, I'll add that to similar class as well

File: android/quest/src/main/java/org/smartregister/fhircore/quest/data/task/PatientTaskRepository.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (pld) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/data/task/PatientTaskRepository.kt:50]: why can't we do that here now, instead of hard coding the ID?

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/PatientTaskFragment.kt [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+      modifier = Modifier,
+      clickListener = { listenerIntent, data -> onItemClicked(listenerIntent, data) }
+    )
+  }
+
+  override fun onItemClicked(listenerIntent: ListenerIntent, data: PatientTaskItem) {
+    // Nothing to navigate at the moment
+  }
+
+  // todo maybe we need to do a db call
```
[the remark this answers is not in the document]
Comment (FikriMilano) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/PatientTaskFragment.kt:75]: Actually for Task, we are not using any of that filtering right now. 
Should we perhaps remove it if that's the case? we can add it once we truly need filtering e.g task searching and overdue filtering.

Sorry, that db thing isn't actually my idea.
@maimoonak could you add some context on the db part?

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/PatientTaskFragment.kt (hunk 1 above)
Comment (FikriMilano) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/PatientTaskFragment.kt:94]: Yeah, I think we can do something similar

File: android/quest/src/main/java/org/smartregister/fhircore/quest/data/task/PatientTaskRepository.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (FikriMilano) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/data/task/PatientTaskRepository.kt:50]: The feature that's supposed to provide the ID is not done yet #857

[the remark this answers is not in the document]
Comment (pld) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/data/task/PatientTaskRepository.kt:50]: oooh I see, reading through that issue, currently we have the keycloak UUID for the logged in user, but not the FHIR ID, is that correct?

[the remark this answers is not in the document]
Comment (pld) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/data/task/PatientTaskRepository.kt:50]: Or asked another way, what is the ID we're passing here vs the IDs we currently have for the logged in user?

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/PatientTaskFragment.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (pld) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/PatientTaskFragment.kt:75]: yea agree, if we're not using it remove for now

File: android/quest/src/main/java/org/smartregister/fhircore/quest/data/task/PatientTaskRepository.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (ellykits) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/data/task/PatientTaskRepository.kt:50]: The ID required on the search API  **Practitioner** resource ID not the keycloak UUID.

[the remark this answers is not in the document]
Comment (ellykits) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/data/task/PatientTaskRepository.kt:50]: And yes we currently have access to the Keycloak UUID.

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/PatientTaskFragment.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (FikriMilano) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/PatientTaskFragment.kt:75]: okay

File: android/quest/src/main/java/org/smartregister/fhircore/quest/data/task/PatientTaskRepository.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (FikriMilano) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/data/task/PatientTaskRepository.kt:50]: > ok got it, thanks. @FikriMilano can you create a new issue noting these lines and the change to make when #857 is complete, and then here and below on line 62, add a todo line referencing the new issue

Awesome, you're welcome. Yeah let's do that also

File: android/quest/src/main/java/org/smartregister/fhircore/quest/util/SearchExtension.kt [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+ * limitations under the License.
+ */
+
+package org.smartregister.fhircore.quest.util
+
+import ca.uhn.fhir.rest.gclient.ReferenceClientParam
+import com.google.android.fhir.search.Search
+import org.hl7.fhir.r4.model.ResourceType
+
+fun Search.filterByPractitioner(reference: ReferenceClientParam, practitionerId: String) {
```
Comment (RaaziaTarique) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/util/SearchExtension.kt:23]: We can generalise Search.filterByPatient method of engine Search extension to Search.filterByResourceType by passing ResourceType name as a parameter and use that here as well .(https://github.com/opensrp/fhircore/blob/main/android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/SearchExtension.kt)

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/component/PatientTaskListItem.kt [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+
+@Composable
+fun TaskLabel(
+  patientItem: PatientTaskItem,
+  clickListener: (PatientTaskListenerIntent, PatientTaskItem) -> Unit,
+  modifier: Modifier
+) {
+
+  val textColor =
+    when (patientItem.overdue) {
```
Comment (RaaziaTarique) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/component/PatientTaskListItem.kt:175]: As patientItem.overdue is a binary condition so you use if here.

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/component/PatientTaskListItem.kt [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+      false -> BlueTextColor
+      true -> OverdueDarkRedColor
+    }
+
+  val bgColor =
+    when (patientItem.overdue) {
```
Comment (RaaziaTarique) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/component/PatientTaskListItem.kt:181]: same here.

File: android/quest/src/main/java/org/smartregister/fhircore/quest/util/SearchExtension.kt (hunk 9 above)
Comment (FikriMilano) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/util/SearchExtension.kt:23]: Yeah sure, I think it's better that way

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/component/PatientTaskListItem.kt (hunk 11 above)
Comment (FikriMilano) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/component/PatientTaskListItem.kt:181]: sure

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/component/PatientTaskListItem.kt [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+@Composable
+fun TaskIcon(
+  patientItem: PatientTaskItem,
+  clickListener: (PatientTaskListenerIntent, PatientTaskItem) -> Unit,
+  modifier: Modifier
+) {
+  val iconId =
+    when (patientItem.overdue) {
```
Comment (RaaziaTarique) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/component/PatientTaskListItem.kt:160]: Use if here too.

Comment (FikriMilano) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/ui/task/component/PatientTaskListItem.kt:160]: Thanks, missed that one

**Review by RaaziaTarique — DISMISSED** [review]

**Review by pld — APPROVED** [review]

**Review by pld — APPROVED** [review]

**Review by pld — APPROVED** [review]
